### PR TITLE
perf: use pipeline for redis rate limiter

### PR DESCRIPTION
**Pull request by anonrig**
Currently, `RedisRateLimiter.is_limited_with_value` takes 8% of an operation such as `/api/0/projects`. This pull request is an attempt to improve the performance.

Instead of making 2 different Redis calls, using `multi` might improve the performance of the rate limiter.

Screenshot 2024-02-01 at 6 02 50 PM

### CODE REVIEW

**Review by volokluev — APPROVED** [review]

File: src/sentry/ratelimits/redis.py
Comment (vartec) [inline src/sentry/ratelimits/redis.py:112]: why the change from `expire` to `execute_command("EXPIRE"`?

File: src/sentry/ratelimits/redis.py
Comment (vartec) [inline src/sentry/ratelimits/redis.py:110]: _MULTI_ makes it a transaction, if the goal of this PR is only to improve the performance, then pipeline w/o a transaction will be enough.

Comment (wedamija) [inline src/sentry/ratelimits/redis.py:110]: It seems like the pipeline is the more important part here, rather than the multi.

I'm not sure how well `.multi` works with `redis-py-cluster`. And I think in CI we're not actually using redis cluster at all, so if it doesn't work, it might just break in production.

**Review by vartec — APPROVED** [review]

File: src/sentry/ratelimits/redis.py
Comment (wedamija) [reply src/sentry/ratelimits/redis.py:110]: Looking at the code, I think this will actually break https://github.com/Grokzen/redis-py-cluster/blob/8a8102a9d758d61a7ec1e2ac9050fcd34029ff3f/rediscluster/pipeline.py#L281

**Review by wedamija — CHANGES_REQUESTED** [review]
Just blocking since I don't think it should merge as is

File: src/sentry/ratelimits/redis.py
Comment (anonrig) [reply src/sentry/ratelimits/redis.py:110]: It seems redis-py-cluster is archived. Maybe we should move to a more maintained solution.

Comment (wedamija) [reply src/sentry/ratelimits/redis.py:110]: Yeah... Seems bad that we're using something unmaintained. I think swapping it out is probably pretty scary too

Comment (anonrig) [reply src/sentry/ratelimits/redis.py:110]: You chose the correct words to motivate me. Thank you! I'll look into it.

**Comment by kibertoad** [comment]
Are you planning to benchmark it?

File: src/sentry/ratelimits/redis.py
Comment (beezz) [reply src/sentry/ratelimits/redis.py:110]: @wedamija we have some preparation for running redis-cluster in dev/ci but no test is using it afaik
https://github.com/getsentry/sentry/pull/49606/files#diff-94dc6dcb2c82ecd1b4fab0edee177a697e821f1340bd6710392b4095880a0e87R60

Comment (anonrig) [reply src/sentry/ratelimits/redis.py:110]: I'm working on removing `redis-py-cluster` at: https://github.com/getsentry/sentry-redis-tools/pull/9

File: src/sentry/ratelimits/redis.py
Comment (anonrig) [reply src/sentry/ratelimits/redis.py:112]: `.expire()` function wasn't implemented for pipeline, but I'm working on a new PR to remove the dependency first: https://github.com/getsentry/sentry-redis-tools/pull/9

Comment (anonrig) [reply src/sentry/ratelimits/redis.py]: Since we updated Redis, we now have `.expire()` command. Resolving this comment. Thank you for reviewing.

File: src/sentry/ratelimits/redis.py
Comment (anonrig) [reply src/sentry/ratelimits/redis.py:111]: I've updated the implementation a little bit, but I left the multi since it provides transactions and reverts expiration if increment fails.

Comment (anonrig) [reply src/sentry/ratelimits/redis.py:111]: Two weeks have passed, and finally we've updated Redis to v4. https://github.com/getsentry/sentry/pull/65282

**Comment by anonrig** [comment]
> Just blocking since I don't think it should merge as is

@wedamija I've updated the implementation since Redis v4 has landed in `master`. Ref: https://github.com/getsentry/sentry/pull/65282

> Are you planning to benchmark it?

@kibertoad I can share the Sentry performance product output once this lands to production.

File: src/sentry/ratelimits/redis.py
Comment (wedamija) [inline src/sentry/ratelimits/redis.py:111]: `multi` is not supported in redis-py either https://github.com/redis/redis-py/blob/master/redis/cluster.py#L2275-L2277

I think it's fine to rely on the pipeline, which is supported https://github.com/redis/redis-py/blob/master/redis/cluster.py#L785

**Review by loewenheim — APPROVED** [review]

File: src/sentry/ratelimits/redis.py
Comment (anonrig) [reply src/sentry/ratelimits/redis.py:111]: TIL

Comment (anonrig) [reply src/sentry/ratelimits/redis.py]: Thank you @wedamija. I've force-pushed and removed multi

**Review by untitaker — APPROVED** [review]
this looks fine but i would let the redis upgrade bake until monday. i was about to comment, i think the use of `pipe.multi()` seems to invoke a lot more complicated python code that probably only slows you down

**Review by wedamija — APPROVED** [review]
Thanks! I'm interested to see how much this improves performance

**Comment by anonrig** [comment]
> this looks fine but i would let the redis upgrade bake until monday. i was about to comment, i think the use of `pipe.multi()` seems to invoke a lot more complicated python code that probably only slows you down

I'll wait until Monday to release this. Removed multi as well. Thanks for the review @untitaker.

**Comment by anonrig** [comment]
Folks, I've added an integration test that runs on an actual redis cluster. This is ready to land. Appreciate it if you could review it regardless. @untitaker @wedamija @loewenheim @volokluev

**Review by volokluev — APPROVED** [review]

**Review by wedamija — APPROVED** [review]

**Review by Swatinem — APPROVED** [review]

**Review by mitsuhiko — APPROVED** [review]

**Pull request merged**
